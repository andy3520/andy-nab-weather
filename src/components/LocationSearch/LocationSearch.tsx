import debounce from 'lodash.debounce';
import React, { FormEvent } from 'react';
import { ReactComponent as SearchIcon } from './search.svg';

export type SearchFunc = (event: React.FormEvent<HTMLInputElement>) => void;
export interface ILocationInputProps {
  onSearch: SearchFunc;
}
const LocationInput: React.FC<ILocationInputProps> = ({ onSearch }) => {
  const [inputEvent, setInputEvent] = React.useState<
    FormEvent<HTMLInputElement>
  >();

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debouncedSearch = React.useCallback(
    debounce((event) => {
      onSearch(event);
    }, 500),
    []
  );

  React.useEffect(() => {
    if (inputEvent) {
      debouncedSearch(inputEvent);
    }
  }, [debouncedSearch, inputEvent]);

  return (
    <div className="w-full max-w-lg lg:max-w-xs">
      <label htmlFor="search" className="sr-only">
        Search
      </label>
      <div className="relative">
        <div className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
          <SearchIcon className="w-5 h-5 text-gray-400 " />
        </div>
        <input
          id="search"
          name="search"
          className="block w-full py-2 pl-10 pr-3 leading-5 placeholder-gray-500 bg-white border border-gray-300 rounded-md focus:outline-none focus:placeholder-gray-400 focus:ring-1 focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
          placeholder="City name"
          type="search"
          onInput={setInputEvent}
        />
      </div>
    </div>
  );
};

export default LocationInput;
