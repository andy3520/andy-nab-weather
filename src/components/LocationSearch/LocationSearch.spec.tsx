import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import { fireEvent, render, waitFor } from '@testing-library/react';
import * as sinon from '@sinonjs/fake-timers';
import LocationSearch, { ELocationSearchTestId } from './LocationSearch';

describe('LocationSearch', () => {
  const onSearchMock = jest.fn();
  // const clock = sinon.
  test('should match snapshot', () => {
    const { asFragment } = render(<LocationSearch onSearch={onSearchMock} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render search input', () => {
    const { getByTestId } = render(<LocationSearch onSearch={onSearchMock} />);

    const INPUT = getByTestId(ELocationSearchTestId.INPUT);
    const ICON = getByTestId(ELocationSearchTestId.ICON);

    expect(ICON).toBeTruthy();
    expect(INPUT).toBeTruthy();
  });

  test('should debounced input callback 500ms', async () => {
    const { getByTestId, getByDisplayValue } = render(
      <LocationSearch onSearch={onSearchMock} />
    );

    const INPUT = getByTestId(ELocationSearchTestId.INPUT);
    const inputValues = ['La', 'ondon'];

    userEvent.type(INPUT, inputValues[0]);
    userEvent.type(INPUT, `{backspace}${inputValues[1]}`);
    expect(getByDisplayValue('London')).toBeTruthy();

    await waitFor(() => expect(onSearchMock).toBeCalledTimes(1));
  });
});
