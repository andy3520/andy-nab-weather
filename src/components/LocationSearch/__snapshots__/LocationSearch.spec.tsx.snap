// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSearch should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full max-w-lg lg:max-w-xs"
  >
    <label
      class="sr-only"
      for="search"
    >
      Search
    </label>
    <div
      class="relative"
    >
      <div
        class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none"
      >
        <svg
          class="w-5 h-5 text-gray-400"
          data-testid="LocationSearch_ICON"
        >
          search.svg
        </svg>
      </div>
      <input
        class="block w-full py-2 pl-10 pr-3 leading-5 placeholder-gray-500 bg-white border border-gray-300 rounded-md focus:outline-none focus:placeholder-gray-400 focus:ring-1 focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
        data-testid="LocationSearch_INPUT"
        id="search"
        name="search"
        placeholder="City name"
        type="search"
      />
    </div>
  </div>
</DocumentFragment>
`;
